runtime/lua/vim/treesitter/query_linter.lua

```diff
@@ -0,0 +1,146 @@
+local namespace = vim.api.nvim_create_namespace('vim.treesitter.query_linter')
+-- those node names exist for every language
+local MAGIC_NODE_NAMES = { '_', 'ERROR' }
+
+local M = {}
+local lints = {}
+
+local function show_lints(buf, lints)
+  local diagnostics = vim.tbl_map(function(lint)
+    return {
+      lnum = lint.range[1],
+      end_lnum = lint.range[3],
+      col = lint.range[2],
+      end_col = lint.range[4],
+      severity = vim.diagnostic.ERROR,
+      message = lint.message,
+    }
+  end, lints)
+  vim.diagnostic.set(namespace, buf, diagnostics)
+end
+
+local function add_lint_for_node(node, buf, error_type, complete_message)
+  local node_text = vim.treesitter.get_node_text(node, buf):gsub('\n', ' ')
+  local error_text = complete_message or error_type .. ': ' .. node_text
+  local error_range = { node:range() }
+  table.insert(
+    lints[buf],
+    { type = error_type, range = error_range, message = error_text, node_text = node_text }
+  )
+end
+
+local function table_contains(predicate, table)
+  for _, elt in pairs(table) do
+    if predicate(elt) then
+      return true
+    end
+  end
+  return false
+end
+
+local function guess_query_lang(buf, _)
+  local filename = vim.api.nvim_buf_get_name(buf)
+  if filename ~= '' then
+    local ok, query_lang = pcall(vim.fn.fnamemodify, filename, ':p:h:t')
+    return ok and query_lang
+  end
+end
+
+local function normalize_opts(buf, opts)
+  opts = opts or {}
+  opts.langs = opts.langs or guess_query_lang
+  if type(opts.langs) == 'function' then
+    opts.langs = opts.langs(buf, opts)
+  end
+  if type(opts.langs) ~= 'table' then
+    opts.langs = { opts.langs }
+  end
+  opts.lints = vim.tbl_extend('keep', opts.lints or {}, {
+    top_level_query = true,
+    fields = true,
+    nodes = true,
+    syntax_error = true,
+  })
+  return opts
+end
+
+local lint_query = [[
+  (program [(named_node) (list) (grouping)] @toplevel-query)
+  (named_node
+    name: _ @named_node)
+  (anonymous_node
+    name: _ @anonymous_node)
+  (field_definition
+    name: (identifier) @field)
+  (ERROR) @error
+]]
```

Comment:
this is a actual query file in nvim-treesitter/playground thought it is merely just a private implementation: https://github.com/nvim-treesitter/playground/blob/master/queries/query/query-linter-queries.scm

Reply:
this file could maybe be generalized to a more general type of treesitter-based linters:
- have a `lint.scm` that define name lint regions
- map each named lint region to check defined by config